### ISSUE TITLE
LaTeX : Missing fontspec for xelatex compilation.

### ISSUE THREAD

**Issue by kpym** [sphinx-doc/sphinx#2727]
The produced LaTeX file doesn't contains `fontspec` package and when it contains accent characters (like `è` for example) when compiled with `xelatex` this characters are missing.

We can see in the produced LaTeX file the following code: 

```
\ifPDFTeX
  \usepackage[utf8]{inputenc}
\fi
```

that could be competed to: 

```
\ifPDFTeX
  % --------------- PDFLaTeX ou LaTeX 
  \usepackage[utf8]{inputenc}
\else
  % --------------- XeLaTeX ou LuaLaTeX
  \usepackage{fontspec}
\fi 
```

**Comment by jfbu** [comment]
Although I agree in principle there are a number of things to consider
-  If the `è` is missing, it means other things where modified. With default `conf.py` this 

``` rst
*****
Hello
*****

éèàç
```

with `make xelatexpdf` has no problem (tested on stable and master branches). But I agree that with some other characters like `«` and `»` you need extra set-up. And loading `fontspec` indirectly provides this.
- the code you refer to is inserted by `'inputenc'` key. This is not good location for  `fontspec`loading.
- `fontspec` makes primarily sense for using OpenType fonts. Hence user overriding default `\usepackage{times)` from `'fontpkg'` key may also be expected to do the right thing with `'fontenc'` key. (see next item)
-  there is a `'fontenc'` key whose default is `'\\usepackage[T1]{fontenc}'`. This is better location for a conditional test of engine.
- one should also worry about `platex/dvipdfmx` road (Japanese)
- loading `fontspec` has an impact on math fonts. Sometimes one needs to load it with `no-math` option.

I think it makes sense that default font configuration should be modified if XeLaTeX/LuaLaTeX is detected and `\usepackage{times}` replaced by suitable `\setmainfont{TeX Gyre Termes}`. Unfortunately this does not work out-of-the-box on all computers. It does not work on Mac OS X without extra set-up (which is easy but not too well known and not explained in either fontspec manual or elsewhere; one should setup symlinks in one's `~/Library/Fonts`, but this impacts LuaLaTeX  (its font database has many duplicates) unless one is using latest TL2016 release) and one must access the font by filename and it becomes cumbersome.

From the fontspec manual:

```
Fonts selected by filename must include bold and italic variants explicitly.
\setmainfont{texgyrepagella-regular.otf}[
BoldFont = texgyrepagella-bold.otf ,
ItalicFont = texgyrepagella-italic.otf ,
BoldItalicFont = texgyrepagella-bolditalic.otf ]
```

Naturally if this is done by latex writer, being cumbersome is not so annoying ...

There has been some previous discussion on these matters. See #2513 for example.

**Comment by jfbu** [comment]
Another thing too add is that at some point in the past the TeX Gyre font filenames have changed. It used to be something like ``tgtermes-regular.otf``. But perhaps one can expect from users wishing to use XeLaTeX and particularly those using LuaLaTeX that they uptodate TeX installations.

**Comment by kpym** [comment]
If the problem is the existence of some fonts on some operating system, it is probably possible to include this fonts in the sphinx distribution ?
Or you can simply put in the requirement of the LaTeX distribution the availability of this fonts. Something like "To be able to compile with Xe/LuaLaTeX you need to install ...".

**Comment by jfbu** [comment]
It is not so much the existence but the fact that for compatibility with all OSs we should use `\setmainfont` in a certain way. But I should have not introduced these peripheral worries in the discussion.

For some years now `TeX Gyre` family of fonts is available and available  from TeXLive/MikTeX. I don't know the status on Linux distros. Again people wishing to use XeLaTeX/LuaLaTeX may be expected to have the corresponding minimal state of affairs.

Termes for times, Heros for helvetica, Chorus for courier.

**Comment by jfbu** [comment]
Hi again, you can already use some set-up like this

``` python
latex_elements = {#

    # The paper size ('letterpaper' or 'a4paper'). 
    'papersize': 'a4paper',

    # The font size ('10pt', '11pt' or '12pt').
    'pointsize': '11pt',

    'inputenc': '',

    'fontenc': '',

    # perhaps some custom Babel or Polyglossia thing
    # 'babel': '\\usepackage{babel}',

    'fontpkg': 
u'''\
\\usepackage{mathtools}
\\usepackage[no-math]{fontspec}%
\\usepackage{unicode-math}
\\setmainfont{Times New Roman}
% more font set up
\\setmathfont[Extension = .otf,BoldFont = *bold]{xits-math}
''',

    # Additional stuff for the LaTeX preamble.
    # 'preamble': '',
}
```

and `make latexpdf PDFLATEX=xelatex`. Future Sphinx-1.5 will have `make xelatexpdf` and `make lualatexpdf`.

Nota Bene: as pointed out by @tk0miya  this is still in discussion at #2513. Hence the make targets may change until release.

**Comment by kpym** [comment]
> Future Sphinx-1.5 will have make xelatexpdf and make lualatexpdf.

This is fantastic news ! Thnx.

If I use your setup it will be compatible only with Xe/LuaLaTeX, but my question is how to produce code that can be compiled with PDFLaTeX and Xe/LuaLaTeX. Can I put `\ifPDFTeX` inside `fontpkg` for example ?

**Comment by jfbu** [comment]
Yes you can. Sphinx loads package iftex since fc3bfbca (1.4.4) hence you have `\ifXeTeX`,  `\ifLuaTeX` conditionals.

By default 

``` python
        'inputenc':        ('\\ifPDFTeX\n'
                            '  \\usepackage[utf8]{inputenc}\n'
                            '\\fi'),
```

this gives you model to do other conditionals in your own set-up.

**Comment by jfbu** [comment]
As per the new make targets it was added at cf3b50d3b9, closing #894

**Comment by kpym** [comment]
Ok. Probably you can close this issue in this case.

**Comment by tk0miya** [comment]
FYI: We are discussing about a new config value for LaTeX engine at #2513.
So the make targets might be changed at release (because v1.5 is still in development!).

**Comment by tk0miya** [comment]
Closed.
Thank you for proposal.

**Comment by jfbu** [comment]
@kpym some of what I said above is obsolete. The `iftex` package is not automatically loaded anymore, as the new `latex_engine` config for `conf.py` has been added at ead753e (PR #2774, merged into 1.5a1). You can check its documentation at http://www.sphinx-doc.org/en/master/config.html#options-for-latex-output and there is an example at http://www.sphinx-doc.org/en/master/latex.html
